Merge isotropic poroelasticity material model

rwalkerlewis opened the pull request.

baagaard-usgs commented:
This PR appears to not be in sync with the current `knepley/feature-petsc-fe`. It removes changes we have made to `knepley/feature-petsc-fe`. Rebasing against `knepley/feature-petsc-fe` will fix this.

rwalkerlewis commented:
Thought I did, but will try again.

On Fri, Oct 16, 2020, 14:43 Brad Aagaard <notifications@github.com> wrote:

> This PR appears to not be in sync with the current
> knepley/feature-petsc-fe. It removes changes we have made to
> knepley/feature-petsc-fe. Rebasing against knepley/feature-petsc-fe will
> fix this.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/geodynamics/pylith/pull/215#issuecomment-710501240>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKJY33SMGUW4Q7KAZBH46PDSLCO45ANCNFSM4STXW5MA>
> .
>

baagaard-usgs commented:
You need to do a force push. `git push --force` after rebasing.

knepley commented:
@baagaard @willic3 I have gone over this for style and obvious things. I just cleaned out some accidental changes. Can you look it over to make sure things are being done in line with the overall design?

knepley commented:
@baagaard Right now, the builds fail due to the PETSc version. I think we need to merge the other PR first.

baagaard-usgs commented:
## Comments

### `libsrc/pylith/bc/NeumannTimeDependent.cc`

* Shouldn't `getPressureScale()` be `getDensityScale()` in [NeumannTimeDependent.cc](https://github.com/geodynamics/pylith/pull/215/files#diff-ab25adfd2d30a9055186257ad938a190057ee9832350337567667dfc39e73e75)

### `libsrc/pylith/fekernels/IsotropicLinearPoroelasticity.cc`
1. Please follow the naming convention for kernel functions (all lower case with underscores with minimal abbreviations). 
2. Declare variables in `for` loops within the loop. `for (PetscInt d=0; ...)`
3. Can you improve the name of the kernels for the MMS tests to something more understandable rather than `_linear_` and `_trig_`? Even just adding `mmstest` to the name would make it better.
4. Please use `sin()`, etc instead of `PetscSinReal()`. We don't use `PetscSinReal()` anywhere else.
5. Please resolve `const PylithInt i_poro_pres = 1; ///SHOULDN'T THIS BE EQUAL TO 1 ??? (JOSIMAR). If it is correct, then delete the comment.
6. Please use the following style to limit line length to less than 120 characters:
```
const PylithInt aOffMean_x[5] = {
    aOff_x[i_undrainedBulkModulus],
    aOff_x[i_biotCoefficient],
    aOff_x[i_biotModulus],
    aOff_x[i_rstress],
    aOff_x[i_rstrain]
};
```
7. Make kernels dimension specific if there are different for 2D and 3D. That is, make a plane strain version and a 3D version. The desire is to remove as many `if` statements as possible.

To be continued...

`

baagaard-usgs commented:
@rwalkerlewis @knepley What is going on? I started reviewing Matt's update (I got through about 18 files) and it had 1 commit of 80 files. This one has 81 files and 365 commits, most of which have useless commit messages.

rwalkerlewis commented:
I pushed by accident and am trying to find how to revert now.

On Mon, Oct 19, 2020 at 9:09 PM Brad Aagaard <notifications@github.com>
wrote:

> @rwalkerlewis <https://github.com/rwalkerlewis> @knepley
> <https://github.com/knepley> What is going on? I started reviewing Matt's
> update (I got through about 18 files) and it had 1 commit of 80 files. This
> one has 81 files and 365 commits, most of which have useless commit
> messages.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geodynamics/pylith/pull/215#issuecomment-712542526>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKJY33QSI2EB34LDO5A2MMTSLTWPDANCNFSM4STXW5MA>
> .
>


-- 
La presente comunicación tiene carácter confidencial y es para el exclusivo
uso del destinatario indicado en la misma. Si Ud. no es el destinatario
indicado, le informamos que cualquier forma de distribución, reproducción o
uso de esta comunicación y/o de la información contenida en la misma están
estrictamente prohibidos por la ley. Si Ud. ha recibido esta comunicación
por error, por favor, notifíquelo inmediatamente al remitente contestando a
este mensaje y proceda a continuación a destruirlo. Gracias por su
colaboración.

This communication contains confidential information. It is for the
exclusive use of the intended addressee. If you are not the intended
addressee, please note that any form of distribution, copying or use of
this communication or the information in it is strictly prohibited by law.
If you have received this communication in error, please immediately notify
the sender by reply e-mail and destroy this message. Thank you for your
cooperation.

baagaard-usgs commented:
@rwalkerlewis You may need Matt to do a force push to overwrite your push.

knepley commented:
Robert, don't worry. I will push tomorrow. I also have all Brad's requested
fixes.

On Mon, Oct 19, 2020 at 10:20 PM Brad Aagaard <notifications@github.com>
wrote:

> @rwalkerlewis <https://github.com/rwalkerlewis> You may need Matt to do a
> force push to overwrite your push.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geodynamics/pylith/pull/215#issuecomment-712545702>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEORCP5XZTQATRPE46UYY3SLTXW3ANCNFSM4STXW5MA>
> .
>


-- 
What most experimenters take for granted before they begin their
experiments is infinitely more interesting than any results to which their
experiments lead.
-- Norbert Wiener

https://www.cse.buffalo.edu/~knepley/ <http://www.cse.buffalo.edu/~knepley/>

baagaard-usgs reviewed (changes requested):
## Review

### General

1. Remove `#include <iostream>` from all files unless required by production code.
2. Please follow the naming convention for kernel functions (all lower case with underscores with limited use of abbreviations). Instead of `QS` and `DYN`, use `_inertia` for the DYN case. 
3. Declare variables in `for` loops within the loop. `for (PetscInt d=0; ...)`
4. Make kernels dimension specific if there are different for 2D and 3D. That is, make a plane strain version and a 3D version. The desire is to remove as many `if` statements as possible. in kernels that ate evaluated in large loops.
5. Is the poroelasticity section of the governing equations in the manual up to date?
6. If porosity is not a parameter (auxiliary subfield), remove all code related to a porosity subfield.

### `libsrc/pylith/bc/NeumannTimeDependent.cc`

* Shouldn't `getPressureScale()` be `getDensityScale()` in [NeumannTimeDependent.cc](https://github.com/geodynamics/pylith/pull/215/files#diff-ab25adfd2d30a9055186257ad938a190057ee9832350337567667dfc39e73e75
 
### `libsrc/pylith/fekernels/IsotropicLinearPoroelasticity.cc`

1. Can you improve the name of the kernels for the MMS tests to something more understandable rather than `_linear_` and `_trig_`? Even just adding `mmstest` to the name would make it better.
2. Please use `sin()`, etc instead of `PetscSinReal()`. We don't use `PetscSinReal()` anywhere else.
3. Please resolve `const PylithInt i_poro_pres = 1; ///SHOULDN'T THIS BE EQUAL TO 1 ??? (JOSIMAR). If it is correct, then delete the comment.
4. Please use the following style to limit line length to less than 120 characters:
```
const PylithInt aOffMean_x[5] = {
    aOff_x[i_undrainedBulkModulus],
    aOff_x[i_biotCoefficient],
    aOff_x[i_biotModulus],
    aOff_x[i_rstress],
    aOff_x[i_rstrain]
};
```

## `libsrc/pylith/fekernels/IsotropicLinearPoroelasticity.hh` 

1. Documentation of required auxiliary subfields (Lines 30-32) seems to need updating. Delete "add" in `addPorosity` etc? Ditto with rheology subfields.
2. In the documentation for the MMS kernels, give specifics (path to files, like `tests/mmstests/poroelasticity/.../MYMMSTEST.cc` of where corresponding tests are.

## `libsrc/pylith/materials/AuxiliaryFactoryPoroelastic.cc`

1. `PYLITH_JOURNAL_DEBUG("isotropicPermeability(void)");` to `PYLITH_JOURNAL_DEBUG("addIsotropicPermeability(void)");` (should match name of method),
2. Ditto for others.
3. Tensor components should be ordered xx, yy, zz, xy, yz, xz for consitency with elasticity. Note that for plane strain we use the first 4 components (xx, yy, zz, xy) while for 3D we use all 6.
4. Why are Young's modulus and Poisson ratio auxiliary subfields? We use the bulk modulus and shear modulus instead.

## `libsrc/pylith/materials/IsotropicLinearPoroelasticity.hh`

Multiple kernels have the exact same description. No two kernels should have the same description.

## `libsrc/pylith/materials/Poroelasticity.hh`

1. `useInertia()` setter/getter methods should be deleted. We now use the formulation, to determine whether inertia is used. See [libsrc/pylith/materials/Elasticity.cc](https://github.com/geodynamics/pylith/blob/knepley/feature-petsc-fe/libsrc/pylith/materials/Elasticity.cc#L141).

## `libsrc/pylith/materials/Poroelasticity.cc`

1. 
    //auxiliaryFactory->addPorosity();      // 0

2. Shouldn't you set kernels for the cases of gravity and/or body forces without source density? Please delete code that is commented out.
3. Why are all the RHS Jacobian functions from the rheology whereas only one residual function is? Do all of them actually depend on the rheology?


## `libsrc/pylith/testing/MMSTest.cc`

`TimeDependent::prestep()` no longer exists. Setting the constrained values is now part of computing the residual.

## `pylith/materials/DerivedSubfieldsPoroelasticity.py`

`DerivedSubfieldsPoroelasticity` seems to be an exact duplicate of `DerivedSubfieldsElasticity`. If they are the same, then simply use `DerivedSubfieldsElasticity`.

## `pylith/materials/Poroelasticity.py`

`useInertia` is no longer a flag for the materials. See comment for `libsrc/pylith/materials/Poroelasticity.hh`.

## `pylith/problems/SolnDispPresVel.py`

Is this set of solution subfields actually used. Isn't trace strain always a solution subfield for poroelasticity? If so, remove this.

## `tests/fullscale/linearporoelasticity/mandel/TestMandel.py`

The output for the boundary conditions is not tested. Please implement these tests.

## `tests/fullscale/linearporoelasticity/mandel/mandel.cfg`

Merge `pylithapp.cfg` and `mandel.cfg` since these apply to all tests in this directory.

The basis order for the solution fields is already set in `pylithapp.cfg`. If you do want to keep them here, then replace
```
[pylithapp.problem.solution.subfields.displacement]
basis_order = 1

[pylithapp.problem.solution.subfields.pressure]
basis_order = 1

[pylithapp.problem.solution.subfields.trace_strain]
basis_order = 1
```

with (matching `pylithapp.cfg`)

```
[pylithapp.problem.solution.subfields]
displacement.basis_order = 1
pressure.basis_order = 1
trace_strain.basis_order = 1
```

## `tests/fullscale/linearporoelasticity/mandel/mandel_gendb.py`

Use a SimpleGridDB. See `./linearelasticity/nofaults-3d/axialdisp_gendb.py`.

## `tests/fullscale/linearporoelasticity/mandel/mandel_quad.cfg`

The progress monitor has been implemented, so uncomment setting the corresponding filename. Ditto for the tri case.

## `tests/fullscale/linearporoelasticity/mandel/mandel_soln.py`

The function `y_pos()` does not seem to be used. If it is used, please rename it to something more meaningful, like `boundary_traction()`.

## `tests/fullscale/linearporoelasticity/mandel/mesh_quad.jou`

Please refactor the journal files to separate the geometry construction and marking boundaries from the meshing. This ensures the geometry and BC for the different cell types remain the same.

## `tests/fullscale/linearporoelasticity/mandel/pylithapp.cfg`

Remove fieldsplit solver settings not used.

## `tests/fullscale/linearporoelasticity/mandel/test.out`

Remove this output file.

## `tests/fullscale/linearporoelasticity/terzaghi/TestTerzaghi.py`

The output for the boundary conditions is not tested. Please implement these tests.

The tri case is commented out. Please implement if possible.

## `tests/fullscale/linearporoelasticity/terzaghi/mesh_quad.jou`

Please refactor the journal files to separate the geometry construction and marking boundaries from the meshing. This ensures the geometry and BC for the different cell types remain the same.

## `tests/fullscale/linearporoelasticity/terzaghi/progress.txt`

Remove this output file.

## `tests/fullscale/linearporoelasticity/terzaghi/pylithapp.cfg`

Remove the `FieldFilterProject` that is commented out.

Remove fieldsplit solver settings not used.

Merge `pylithapp.cfg` and `terzaghi.cfg` since these apply to all tests in this directory.

## `tests/fullscale/linearporoelasticity/terzaghi/terzaghi.cfg`

Condense specification of basis order as is done in `pylithapp.cfg`/

Use `defaults.output_field_filter = pylith.meshio.FieldFilterProject` instead of setting `FieldFilterProject` for all observers.

You shouldn't have to set the `data_fields` to all solution subfields. This should be the default.

## `tests/fullscale/linearporoelasticity/terzaghi/terzaghi_gendb.py`

Use a SimpleGridDB. See `./linearelasticity/nofaults-3d/axialdisp_gendb.py`.

Remove commented code.

## `tests/fullscale/linearporoelasticity/terzaghi/terzaghi_quad.cfg`

The progress monitor has been implemented, so uncomment setting the corresponding filename. Ditto for the tri case.

## `tests/fullscale/linearporoelasticity/terzaghi/terzaghi_soln.py`

Remove commented code.

## `tests/fullscale/linearporoelasticity/terzaghi/test.out`

Remove this output file.

baagaard-usgs commented:
For ease of reviewing, I suggest that we start with Matt's single commit and changes based on my review comments be in another commit or small series of commits.

knepley commented:
> ## Review
> ### General
> 1. Remove `#include <iostream>` from all files unless required by production code.
> 2. Please follow the naming convention for kernel functions (all lower case with underscores with limited use of abbreviations). Instead of `QS` and `DYN`, use `_inertia` for the DYN case.
> 3. Declare variables in `for` loops within the loop. `for (PetscInt d=0; ...)`
> 4. Make kernels dimension specific if there are different for 2D and 3D. That is, make a plane strain version and a 3D version. The desire is to remove as many `if` statements as possible. in kernels that ate evaluated in large loops.
> 5. Is the poroelasticity section of the governing equations in the manual up to date?
> 6. If porosity is not a parameter (auxiliary subfield), remove all code related to a porosity subfield.
> 
> ### `libsrc/pylith/bc/NeumannTimeDependent.cc`
> * Shouldn't `getPressureScale()` be `getDensityScale()` in [NeumannTimeDependent.cc](https://github.com/geodynamics/pylith/pull/215/files#diff-ab25adfd2d30a9055186257ad938a190057ee9832350337567667dfc39e73e75

Fixed

> ### `libsrc/pylith/fekernels/IsotropicLinearPoroelasticity.cc`
> 1. Can you improve the name of the kernels for the MMS tests to something more understandable rather than `_linear_` and `_trig_`? Even just adding `mmstest` to the name would make it better.
> 2. Please use `sin()`, etc instead of `PetscSinReal()`. We don't use `PetscSinReal()` anywhere else.
> 3. Please resolve `const PylithInt i_poro_pres = 1; ///SHOULDN'T THIS BE EQUAL TO 1 ??? (JOSIMAR). If it is correct, then delete the comment.
> 4. Please use the following style to limit line length to less than 120 characters:
> 
> ```
> const PylithInt aOffMean_x[5] = {
>     aOff_x[i_undrainedBulkModulus],
>     aOff_x[i_biotCoefficient],
>     aOff_x[i_biotModulus],
>     aOff_x[i_rstress],
>     aOff_x[i_rstrain]
> };
> ```
> 
> ## `libsrc/pylith/fekernels/IsotropicLinearPoroelasticity.hh`
> 1. Documentation of required auxiliary subfields (Lines 30-32) seems to need updating. Delete "add" in `addPorosity` etc? Ditto with rheology subfields.
> 2. In the documentation for the MMS kernels, give specifics (path to files, like `tests/mmstests/poroelasticity/.../MYMMSTEST.cc` of where corresponding tests are.
> 
> ## `libsrc/pylith/materials/AuxiliaryFactoryPoroelastic.cc`
> 1. `PYLITH_JOURNAL_DEBUG("isotropicPermeability(void)");` to `PYLITH_JOURNAL_DEBUG("addIsotropicPermeability(void)");` (should match name of method),
> 2. Ditto for others.
> 3. Tensor components should be ordered xx, yy, zz, xy, yz, xz for consitency with elasticity. Note that for plane strain we use the first 4 components (xx, yy, zz, xy) while for 3D we use all 6.
> 4. Why are Young's modulus and Poisson ratio auxiliary subfields? We use the bulk modulus and shear modulus instead.
> 
> ## `libsrc/pylith/materials/IsotropicLinearPoroelasticity.hh`
> Multiple kernels have the exact same description. No two kernels should have the same description.
> 
> ## `libsrc/pylith/materials/Poroelasticity.hh`
> 1. `useInertia()` setter/getter methods should be deleted. We now use the formulation, to determine whether inertia is used. See [libsrc/pylith/materials/Elasticity.cc](https://github.com/geodynamics/pylith/blob/knepley/feature-petsc-fe/libsrc/pylith/materials/Elasticity.cc#L141).
> 
> ## `libsrc/pylith/materials/Poroelasticity.cc`
> 1. //auxiliaryFactory->addPorosity();      // 0
> 2. Shouldn't you set kernels for the cases of gravity and/or body forces without source density? Please delete code that is commented out.
> 3. Why are all the RHS Jacobian functions from the rheology whereas only one residual function is? Do all of them actually depend on the rheology?
> 
> ## `libsrc/pylith/testing/MMSTest.cc`
> `TimeDependent::prestep()` no longer exists. Setting the constrained values is now part of computing the residual.
> 
> ## `pylith/materials/DerivedSubfieldsPoroelasticity.py`
> `DerivedSubfieldsPoroelasticity` seems to be an exact duplicate of `DerivedSubfieldsElasticity`. If they are the same, then simply use `DerivedSubfieldsElasticity`.
> 
> ## `pylith/materials/Poroelasticity.py`
> `useInertia` is no longer a flag for the materials. See comment for `libsrc/pylith/materials/Poroelasticity.hh`.
> 
> ## `pylith/problems/SolnDispPresVel.py`
> Is this set of solution subfields actually used. Isn't trace strain always a solution subfield for poroelasticity? If so, remove this.
> 
> ## `tests/fullscale/linearporoelasticity/mandel/TestMandel.py`
> The output for the boundary conditions is not tested. Please implement these tests.
> 
> ## `tests/fullscale/linearporoelasticity/mandel/mandel.cfg`
> Merge `pylithapp.cfg` and `mandel.cfg` since these apply to all tests in this directory.
> 
> The basis order for the solution fields is already set in `pylithapp.cfg`. If you do want to keep them here, then replace
> 
> ```
> [pylithapp.problem.solution.subfields.displacement]
> basis_order = 1
> 
> [pylithapp.problem.solution.subfields.pressure]
> basis_order = 1
> 
> [pylithapp.problem.solution.subfields.trace_strain]
> basis_order = 1
> ```
> 
> with (matching `pylithapp.cfg`)
> 
> ```
> [pylithapp.problem.solution.subfields]
> displacement.basis_order = 1
> pressure.basis_order = 1
> trace_strain.basis_order = 1
> ```
> 
> ## `tests/fullscale/linearporoelasticity/mandel/mandel_gendb.py`
> Use a SimpleGridDB. See `./linearelasticity/nofaults-3d/axialdisp_gendb.py`.
> 
> ## `tests/fullscale/linearporoelasticity/mandel/mandel_quad.cfg`
> The progress monitor has been implemented, so uncomment setting the corresponding filename. Ditto for the tri case.
> 
> ## `tests/fullscale/linearporoelasticity/mandel/mandel_soln.py`
> The function `y_pos()` does not seem to be used. If it is used, please rename it to something more meaningful, like `boundary_traction()`.
> 
> ## `tests/fullscale/linearporoelasticity/mandel/mesh_quad.jou`
> Please refactor the journal files to separate the geometry construction and marking boundaries from the meshing. This ensures the geometry and BC for the different cell types remain the same.
> 
> ## `tests/fullscale/linearporoelasticity/mandel/pylithapp.cfg`
> Remove fieldsplit solver settings not used.
> 
> ## `tests/fullscale/linearporoelasticity/mandel/test.out`
> Remove this output file.
> 
> ## `tests/fullscale/linearporoelasticity/terzaghi/TestTerzaghi.py`
> The output for the boundary conditions is not tested. Please implement these tests.
> 
> The tri case is commented out. Please implement if possible.
> 
> ## `tests/fullscale/linearporoelasticity/terzaghi/mesh_quad.jou`
> Please refactor the journal files to separate the geometry construction and marking boundaries from the meshing. This ensures the geometry and BC for the different cell types remain the same.
> 
> ## `tests/fullscale/linearporoelasticity/terzaghi/progress.txt`
> Remove this output file.
> 
> ## `tests/fullscale/linearporoelasticity/terzaghi/pylithapp.cfg`
> Remove the `FieldFilterProject` that is commented out.
> 
> Remove fieldsplit solver settings not used.
> 
> Merge `pylithapp.cfg` and `terzaghi.cfg` since these apply to all tests in this directory.
> 
> ## `tests/fullscale/linearporoelasticity/terzaghi/terzaghi.cfg`
> Condense specification of basis order as is done in `pylithapp.cfg`/
> 
> Use `defaults.output_field_filter = pylith.meshio.FieldFilterProject` instead of setting `FieldFilterProject` for all observers.
> 
> You shouldn't have to set the `data_fields` to all solution subfields. This should be the default.
> 
> ## `tests/fullscale/linearporoelasticity/terzaghi/terzaghi_gendb.py`
> Use a SimpleGridDB. See `./linearelasticity/nofaults-3d/axialdisp_gendb.py`.
> 
> Remove commented code.
> 
> ## `tests/fullscale/linearporoelasticity/terzaghi/terzaghi_quad.cfg`
> The progress monitor has been implemented, so uncomment setting the corresponding filename. Ditto for the tri case.
> 
> ## `tests/fullscale/linearporoelasticity/terzaghi/terzaghi_soln.py`
> Remove commented code.
> 
> ## `tests/fullscale/linearporoelasticity/terzaghi/test.out`
> Remove this output file.

rwalkerlewis commented:
I'm not sure if Dr. Knepley's final push has gone through yet, however I felt it necessary to address what are the major issues from my standpoint:

> > ## Review
> > ### General
> > 
> > 1. Remove `#include <iostream>` from all files unless required by production code.
> > 2. Please follow the naming convention for kernel functions (all lower case with underscores with limited use of abbreviations). Instead of `QS` and `DYN`, use `_inertia` for the DYN case.
> > 3. Declare variables in `for` loops within the loop. `for (PetscInt d=0; ...)`
> > 4. Make kernels dimension specific if there are different for 2D and 3D. That is, make a plane strain version and a 3D version. The desire is to remove as many `if` statements as possible. in kernels that ate evaluated in large loops.
> > 5. Is the poroelasticity section of the governing equations in the manual up to date?
> > 6. If porosity is not a parameter (auxiliary subfield), remove all code related to a porosity subfield.
> 
> I put my fixes in separate ChangeSets so you can see them. They can be squashed when we are done.
> 
> 1. Fixed
> 2. FIxed
> 3. Fixed
> 4. I do not understand. Robert, do you understand?

The IsotropicLinearElasticity case uses distinct plane strain and 3D kernels; is this the form that should be followed? If it is just a matter of eliminating if statements, that can be done, and is done for calculations without reference stress or strain (also cauchy stress calculation uses them, but they are still legacy code.) I would prefer the kernels to be dimensionally agnostic, but will change if necessary,

> 5. I do not think so.
> 6. Robert will have to do this after I push.

Reference to porosity as a parameter exists because it was an input subfield prior to our reorganization around the current input fields. I am of the opinion that it should be re-instituted as an input field, as the current input fields for poroelasticity are not traditional and will likely serve to confuse any end user. Is there a way to have variable field input (i.e. the current variables, or some combination of other variables that are more commonplace,) and just have the code deal with conversion?
> 

The other issues that need addressing don't really require discussion, not before any review, at least.

baagaard-usgs commented:
Yes, follow `IsotropicLinearElasticity` with distinct plane strain and 3D kernels.

> Is there a way to have variable field input (i.e. the current variables, or some combination of other variables that are more commonplace,) and just have the code deal with conversion?

Yes. For elasticity internally we store density, bulk modulus, and shear modulus while the input fields are density, Vp, and Vs. I recommend something similar for poroelasticity. See https://github.com/geodynamics/pylith/blob/knepley/feature-petsc-fe/libsrc/pylith/materials/AuxiliaryFactoryElastic.cc#L67 and https://github.com/geodynamics/pylith/blob/knepley/feature-petsc-fe/libsrc/pylith/materials/Query.cc#L90.

baagaard-usgs commented:
@rwalkerlewis The libtests are failing for this PR (see the checks that failed above and/or run the tests on your machine). All tests pass for the `knepley/feature-petsc-fe` branch (which you should have used to rebase your branch).

## .gitignore

We never build in our source directory (we use separate build directories for multiple configurations, e.g., debugging and optimized versions), so many of the files (`*.Plo`, `*.lo`, `*.o`, `*.Po`, `portinfo`, `libtool`, `*.status`) you have added to `.gitignore` are superfluous. Also the source repository contains some files with extensions that you have incorrectly added to `.gitignore`, such as `*.spatialdb`, `*.exo`, `*.timedb`, `*.h5`, `*.xmf`. Please remove all of these from `.gitignore`.

## tests/fullscale/\*/\*/*gendb.py

For some reason the PR is showing the permissions for these files were changed. They should be 100755 not 100644.

## libsrc/pylith/materials/Poroelasticity.cc

Line 161: Isn't `trace_strain` needed as a subfield whether or not inertia is used?

Lines 331-365: You don't handle any cases in which you have gravity or body forces but no source density. Additionally, it looks like you don't actually include the body force or gravitational body forces in the kernels when they are turned "on". For example, `g0p_sourceDensity_grav` just calls `g0p_source`. It should call a kernel to add the gravitational body forces too.

## pylith/libsrc/pylith/materials/DerivedFactoryPoroelasticity.cc

DerivedSubfieldsPoroelasticity seems to be an exact duplicate of DerivedSubfieldsElasticity. If they are the same, then simply use DerivedSubfieldsElasticity. You removed the corresponding Python file but not the C++ file. Also remove the module file (.i) if it exists.

## tests/fullscale/linearporoelasticity/mandel

Merge `mandel.cfg` into `pylithapp.cfg` since those settings apply to all tests in this directory.

Please refactor the journal files to separate the geometry construction and marking boundaries from the meshing. This ensures the geometry and BC for the different cell types remain the same.

## tests/fullscale/linearporoelasticity/terzaghi

Merge `terzaghi.cfg` into `pylithapp.cfg` since those settings apply to all tests in this directory.

Please refactor the journal files to separate the geometry construction and marking boundaries from the meshing. This ensures the geometry and BC for the different cell types remain the same.

Remove unused code from terzaghi_gendb.py.

rwalkerlewis commented:
> ## tests/fullscale/*/*/*gendb.py
> For some reason the PR is showing the permissions for these files were changed. They should be 100755 not 100644.

I am not sure how his is changed or how to fix this.

baagaard-usgs commented:
To fix the permissions, run `chmod +x $FILE` where $FILE is all of the gendb.py files that have permissions set to 644 (-x).

rwalkerlewis commented:
The CI tests appear to be failing consistently when the MMSTest functions testResidual and testJacobianTaylorSeries are called.
Here is an example of the error output (PETSc related):
`pylith::mmstests::TestIsotropicLinearElasticity2D_BodyForce_QuadQ3::testResidual[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Null argument, when expecting valid pointer
[0]PETSC ERROR: Null Object: Parameter # 1
[0]PETSC ERROR: See https://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[0]PETSC ERROR: Petsc Development GIT revision: v3.14-201-g87ebeec971  GIT Date: 2020-10-22 06:39:33 -0400
[0]PETSC ERROR: /home/pylith-user/build/pylith/tests/mmstests/elasticity/.libs/test_isotropiclinearelasticity2d on a arch-pylith named eecaf18b4b8f by Unknown Wed Dec  2 03:54:13 2020
[0]PETSC ERROR: Configure options --download-chaco=1 --download-ml=1 --download-f2cblaslapack=1 --with-hdf5=1 --with-hdf5-include=/usr/include/hdf5/mpich --with-hdf5-lib=/usr/lib/x86_64-linux-gnu/hdf5/mpich/lib/libhdf5.so --with-zlib=1 --LIBS=-lz --with-debugging=1 --with-fc=0 --with-64-bit-points=1 --with-large-file-io=1 CPPFLAGS=-I/opt/pylith/dependencies/include LDFLAGS=-L/opt/pylith/dependencies/lib CFLAGS="-g -O"
[0]PETSC ERROR: #161 VecAXPBYPCZ() line 680 in /opt/pylith/petsc/src/vec/vec/interface/rvector.c
[0]PETSC ERROR: #162 SNESTSFormFunction_Theta() line 961 in /opt/pylith/petsc/src/ts/impls/implicit/theta/theta.c
[0]PETSC ERROR: #163 SNESTSFormFunction() line 5014 in /opt/pylith/petsc/src/ts/interface/ts.c
[0]PETSC ERROR: #164 SNESComputeFunction() line 2374 in /opt/pylith/petsc/src/snes/interface/snes.c
[0]PETSC ERROR: #165 DMSNESCheckResidual() line 1640 in /opt/pylith/petsc/src/snes/utils/dmplexsnes.c
 : assertion
pylith::mmstests::TestIsotropicLinearElasticity2D_BodyForce_QuadQ3::testJacobianTaylorSeries[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Null argument, when expecting valid pointer
[0]PETSC ERROR: Null Object: Parameter # 4
[0]PETSC ERROR: See https://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[0]PETSC ERROR: Petsc Development GIT revision: v3.14-201-g87ebeec971  GIT Date: 2020-10-22 06:39:33 -0400
[0]PETSC ERROR: /home/pylith-user/build/pylith/tests/mmstests/elasticity/.libs/test_isotropiclinearelasticity2d on a arch-pylith named eecaf18b4b8f by Unknown Wed Dec  2 03:54:13 2020
[0]PETSC ERROR: Configure options --download-chaco=1 --download-ml=1 --download-f2cblaslapack=1 --with-hdf5=1 --with-hdf5-include=/usr/include/hdf5/mpich --with-hdf5-lib=/usr/lib/x86_64-linux-gnu/hdf5/mpich/lib/libhdf5.so --with-zlib=1 --LIBS=-lz --with-debugging=1 --with-fc=0 --with-64-bit-points=1 --with-large-file-io=1 CPPFLAGS=-I/opt/pylith/dependencies/include LDFLAGS=-L/opt/pylith/dependencies/lib CFLAGS="-g -O"
[0]PETSC ERROR: #166 TSComputeIJacobian() line 955 in /opt/pylith/petsc/src/ts/interface/ts.c
[0]PETSC ERROR: #167 SNESTSFormJacobian_Theta() line 989 in /opt/pylith/petsc/src/ts/impls/implicit/theta/theta.c
[0]PETSC ERROR: #168 SNESTSFormJacobian() line 5053 in /opt/pylith/petsc/src/ts/interface/ts.c
[0]PETSC ERROR: #169 SNESComputeJacobian() line 2708 in /opt/pylith/petsc/src/snes/interface/snes.c
[0]PETSC ERROR: #170 DMSNESCheckJacobian() line 1705 in /opt/pylith/petsc/src/snes/utils/dmplexsnes.c
 : assertion`

rwalkerlewis commented:
@knepley How can the aforementioned errors be dealt with? I think it is a petsc issue.

knepley commented:
@rwalkerlewis Tell me how to run the failing test and I can debug things

rwalkerlewis commented:
the test that is failing is test_isotropiclinearelasticity2d, from that which is run under make check

rwalkerlewis commented:
@baagaard-usgs CI is passing, could we review for merge again?

baagaard-usgs reviewed (changes requested):
We are getting very close!

## .gitignore

Remove the following from .gitignore:

```
Makefile
portinfo
libtool
.log
*.txt
*.json
m4
```


## libsrc/pylith/materials/Poroelasticity.cc

You have flags for including body forces and gravitational body forces and include them in the auxiliary field, but you don't set kernels for them. You should be able to make use of the existing kernels for computing these. It is going to be easier to set the kernels for these missing cases rather than trap the errors that exist because these cases are not implemented. As in the previous review comments, you should be considering 7 cases in `_setKernelsRHSResidual` for which options are being used. 

## tests/fullscale/linearporoelasticity/mandel

`TestMandel.py`: Remove commented out code for Neumann BC.

`mesh_tri.jou`: Mesh size is not specified (currently it is commented out and looks to have incorrect value).

Please consolidate `pylithapp.cfg` and `mandel.cfg` as both are used for all simulations in this directory.

## tests/fullscale/linearporoelasticity/terzaghi

`mesh_tri.jou`: Mesh size is not specified (currently it is commented out).

Please consolidate `pylithapp.cfg` and `terzaghi.cfg` as both are used for all simulations in this directory.

## tests/fullscale/viscoelasticity

The permissions for the following files should be 755 (+x) not 644 (-x).

```
tests/fullscale/viscoelasticity/nofaults-2d/axialstrain_genmaxwell_gendb.py
tests/fullscale/viscoelasticity/nofaults-2d/axialstrainrate_genmaxwell_gendb.py
tests/fullscale/viscoelasticity/nofaults-2d/axialtraction_maxwell_gendb.py
tests/fullscale/viscoelasticity/nofaults-3d/axialstrain_genmaxwell_gendb.py
tests/fullscale/viscoelasticity/nofaults-3d/axialstrainrate_genmaxwell_gendb.py
tests/fullscale/viscoelasticity/nofaults-3d/axialtraction_maxwell_gendb.py
```

rwalkerlewis commented:
> We are getting very close!
> 
> ## .gitignore
> Remove the following from .gitignore:
> 
> ```
> Makefile
> portinfo
> libtool
> .log
> *.txt
> *.json
> m4
> ```
> 

Done

> ## libsrc/pylith/materials/Poroelasticity.cc
> You have flags for including body forces and gravitational body forces and include them in the auxiliary field, but you don't set kernels for them. You should be able to make use of the existing kernels for computing these. It is going to be easier to set the kernels for these missing cases rather than trap the errors that exist because these cases are not implemented. As in the previous review comments, you should be considering 7 cases in `_setKernelsRHSResidual` for which options are being used.
> 

The RHS kernels did not need or use all those additional flags. All flags are used within the code.

> ## tests/fullscale/linearporoelasticity/mandel
> `TestMandel.py`: Remove commented out code for Neumann BC.
> 
Done
> `mesh_tri.jou`: Mesh size is not specified (currently it is commented out and looks to have incorrect value).
> 
Done
> Please consolidate `pylithapp.cfg` and `mandel.cfg` as both are used for all simulations in this directory.
> 
Done
> ## tests/fullscale/linearporoelasticity/terzaghi
> `mesh_tri.jou`: Mesh size is not specified (currently it is commented out).
> 
Done
> Please consolidate `pylithapp.cfg` and `terzaghi.cfg` as both are used for all simulations in this directory.
> 
Done
> ## tests/fullscale/viscoelasticity
> The permissions for the following files should be 755 (+x) not 644 (-x).
> 
> ```
> tests/fullscale/viscoelasticity/nofaults-2d/axialstrain_genmaxwell_gendb.py
> tests/fullscale/viscoelasticity/nofaults-2d/axialstrainrate_genmaxwell_gendb.py
> tests/fullscale/viscoelasticity/nofaults-2d/axialtraction_maxwell_gendb.py
> tests/fullscale/viscoelasticity/nofaults-3d/axialstrain_genmaxwell_gendb.py
> tests/fullscale/viscoelasticity/nofaults-3d/axialstrainrate_genmaxwell_gendb.py
> tests/fullscale/viscoelasticity/nofaults-3d/axialtraction_maxwell_gendb.py
> ```
Done

baagaard-usgs commented:
Everything looks good except setting the RHS residual kernels in [`Poroelasticity.cc`](https://github.com/rwalkerlewis/pylith/blob/rlwalker/poro-mms-test/libsrc/pylith/materials/Poroelasticity.cc#L339).

The logic for gravity and body forces being on/off for the g0u and g0v kernels can be encapsulated in the switch statement thereby reducing duplicate code. You also need to account for the differences in offset for the source density auxiliary subfield as a result of gravity and body forces being present or absent.

```
const int bitBodyForce = _useBodyForce ? 0x1 : 0x0;
const int bitGravity = _gravityField ? 0x2 : 0x0;
const int bitSourceDensity = _useSourceDensity ? 0x4 : 0x0;
const int bitUse = bitBodyForce | bitGravity | bitSourceDensity;

switch (bitUse) {
case 0x1:
    g0v = pylith::fekernels::Poroelasticity::g0v_bodyforce;
    break;
case 0x2:
    g0v = pylith::fekernels::Poroelasticity::g0v_gravity;
    break;
case 0x4:
    g0p = pylith::fekernels::Poroelasticity::g0p_sourceDensity; // aOff for sourceDensity is 3
    break;
case 0x3:
    g0v = pylith::fekernels::Poroelasticity::g0v_gravity_bodyforce;
    break;
case 0x5:
    g0v = pylith::fekernels::Poroelasticity::g0v_bodyforce;
    g0p = pylith::fekernels::Poroelasticity::g0p_bodyforce_sourceDensity; // aOff for sourceDensity is 4
    break;
case 0x6:
    g0v = pylith::fekernels::Poroelasticity::g0v_gravity;
    g0p = pylith::fekernels::Poroelasticity::g0p_gravity_sourceDensity;  // aOff for sourceDensity is 4
    break;
case 0x7:
    g0v = pylith::fekernels::Poroelasticity::g0v_gravity_bodyforce;
    g0p = pylith::fekernels::Poroelasticity::g0p_gravity_bodyforce_sourceDensity;  // aOff for sourceDensity is 5
    break;
case 0x0: {
        break;
default:
        PYLITH_COMPONENT_FIREWALL("Unknown case (bitUse=" << bitUse << ") for Poroelasticity RHS residual kernels.");
} // switch
```

baagaard-usgs reviewed (changes requested):
Everything looks good except setting the RHS residual kernels in [`Poroelasticity.cc`](https://github.com/rwalkerlewis/pylith/blob/rlwalker/poro-mms-test/libsrc/pylith/materials/Poroelasticity.cc#L339).

The logic for gravity and body forces being on/off for the g0u and g0v kernels can be encapsulated in the switch statement thereby reducing duplicate code. You also need to account for the differences in offset for the source density auxiliary subfield as a result of gravity and body forces being present or absent.

```
const int bitBodyForce = _useBodyForce ? 0x1 : 0x0;
const int bitGravity = _gravityField ? 0x2 : 0x0;
const int bitSourceDensity = _useSourceDensity ? 0x4 : 0x0;
const int bitUse = bitBodyForce | bitGravity | bitSourceDensity;

switch (bitUse) {
case 0x1:
    g0v = pylith::fekernels::Poroelasticity::g0v_bodyforce;
    break;
case 0x2:
    g0v = pylith::fekernels::Poroelasticity::g0v_gravity;
    break;
case 0x4:
    g0p = pylith::fekernels::Poroelasticity::g0p_sourceDensity; // aOff for sourceDensity is 3
    break;
case 0x3:
    g0v = pylith::fekernels::Poroelasticity::g0v_gravity_bodyforce;
    break;
case 0x5:
    g0v = pylith::fekernels::Poroelasticity::g0v_bodyforce;
    g0p = pylith::fekernels::Poroelasticity::g0p_bodyforce_sourceDensity; // aOff for sourceDensity is 4
    break;
case 0x6:
    g0v = pylith::fekernels::Poroelasticity::g0v_gravity;
    g0p = pylith::fekernels::Poroelasticity::g0p_gravity_sourceDensity;  // aOff for sourceDensity is 4
    break;
case 0x7:
    g0v = pylith::fekernels::Poroelasticity::g0v_gravity_bodyforce;
    g0p = pylith::fekernels::Poroelasticity::g0p_gravity_bodyforce_sourceDensity;  // aOff for sourceDensity is 5
    break;
case 0x0: {
        break;
default:
        PYLITH_COMPONENT_FIREWALL("Unknown case (bitUse=" << bitUse << ") for Poroelasticity RHS residual kernels.");
} // switch
```

baagaard-usgs reviewed (approved):
Looks good!
